api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto

```diff
@@ -21,19 +22,65 @@ message TcpProxy {
-  //  Otherwise, a :ref:`deprecated_v1
-  //  <envoy_api_field_config.filter.network.tcp_proxy.v2.TcpProxy.deprecated_v1>` configuration is
-  //  required to use more complex routing in the interim.
-  //
-  string cluster = 2;
+  // [#not-implemented-hide:] Weighted clusters allow for specification of
+  // multiple upstream clusters along with weights that indicate the
+  // percentage of traffic to be forwarded to each cluster. TCP Proxy
+  // selects an upstream cluster based on the weights.
+  message WeightedCluster {
+    message ClusterWeight {
+      // Name of the upstream cluster. The cluster must exist in the
+      // :ref:`cluster manager configuration <config_cluster_manager>`.
+      string name = 1 [(validate.rules).string.min_bytes = 1];
+      
+      // An integer between 0 and :ref:`total_weight
+      // <envoy_api_field_route.WeightedCluster.total_weight>`. The choice of an upstream
+      // cluster is determined by its weight. The sum of weights across all entries in the
+      // clusters array must add up to the total_weight, which defaults to 100.
+      google.protobuf.UInt32Value weight = 2;
+
+      // [#not-implemented-hide:] Optional endpoint metadata match
+      // criteria. Only endpoints in the upstream cluster with metadata
+      // matching that set in metadata_match will be considered. The filter
+      // name should be specified as *envoy.lb*.
+      core.Metadata metadata_match = 3;
+
+      // [#not-implemented-hide:] The per_filter_config field can be used
+      // to provide weighted cluster-specific configurations for
+      // filters. The key should match the filter name, such as
+      // *envoy.buffer* for the HTTP buffer filter. Use of this field is
+      // filter specific; see the :ref:`HTTP filter documentation
+      // <config_http_filters>` for if and how it is utilized.
+      map<string, google.protobuf.Struct> per_filter_config = 4;
+    }
+
+    // Specifies one or more upstream clusters associated with the TCP proxy.
+    repeated ClusterWeight clusters = 1 [(validate.rules).repeated .min_items = 1];
+
+    // Specifies the total weight across all clusters. The sum of all cluster
+    // weights must equal this value, which must be greater than 0. Defaults
+    // to 100.
+    google.protobuf.UInt32Value total_weight = 3 [(validate.rules).uint32.gte = 1];
+  }
+  
+  oneof cluster_specifier {
+    option (validate.required) = true;
+
+    // The upstream cluster to connect to.
+    string cluster = 2 [(validate.rules).string.min_bytes = 1];
+
+    // [#not-implemented-hide:] Envoy will determine the cluster to route
+    // to based on the SNI presented by the client during the TLS
+    // handshake. If the SNI is empty (as is the case for plaintext TCP
+    // connections) or the referenced cluster does not exist, Envoy will
+    // terminate the connection.
+    google.protobuf.BoolValue use_sni_as_cluster = 10;
```

Comment:
Let's use `server_names` instead of `sni` here, to be consistent with what `FilterChainMatch` has, and put SNI only where it is TLS specific.

Also I suggest put a strong warning here that this is pretty dangerous when used without specifying `server_names` in `FilterChainMatch` and/or RBAC filter in front of it. It will allow client to request arbitrary cluster configured in the Envoy (e.g., Secret Discovery Service).

Reply:
Sure. See Greg’s comments. 

Also, this behavior is already present today with the cluster_header option in http and also the original dst ip from the headers. So we might want to define a separate thing that marks one or more clusters as private or not addressable from filters